### PR TITLE
add notification if tsserver logs are left on for more than 7 days

### LINKED ISSUE #123493: Suggest switching off TSServer Trace Logs if left on for a long time
If TS Server crashes 5 times, VSCode has a nice popup that suggests some action, which includes automatically turning on TSServer Trace Logs. It's very easy to forget that this option is on (or even be unaware that it was switched on at all) and very hard to diagnose the resulting performance problem in the wider IDE ("intellisense gradually gets slower over time" is difficult to diagnose down to an option you didn't know exists and that's cause by accumulating GiB of log files).

I would like that if the TSServer Enable Trace flag has been on for more than 7 days, VSCode once a week suggests turning it off. For the vast majority of users, this is likely to help them. For the small number of people who really do need this flag on all the time, a once a week popup won't hurt (or they can have some option to permanently turn off the reminder).

This would have avoided my month-long headache that eventually culminated in #123446 (and was only resolved when I used WinDirStat to find out why my system drive was full).

See also: https://twitter.com/orta/status/1391757821017010176?s=20

### PR DESCRIPTION
**Pull request by nrayburn-tech**
If the typescript server logs are enabled for more than 7 days, then a notification is displayed with an option to disable.

I can make a few more improvements, but I just wanted to make sure I was making the changes in the appropriate area.

This PR fixes #123493.

### CODE REVIEW

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 1]
```diff
@@ -952,6 +957,47 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 			this.executeWithoutWaitingForResponse('configurePlugin', { pluginName, configuration });
 		}
 	}
+
+	private notifyExtendedLogging(context: vscode.ExtensionContext) {
+		const getLastChange = (): Date | undefined => {
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:962]: There needs to be a way to disable this prompt too. I always keep this setting enabled so I can easily collect logs when something weird happens

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 2]
```diff
@@ -952,6 +957,47 @@ export default class TypeScriptServiceClient extends Disposable implements IType
...
+		const lastLogLevelChange = getLastChange();
+
+		if (logLevel !== TsServerLogLevel.Off && lastLogLevelChange) {
+			const lastChangePlusOneWeek = new Date(lastLogLevelChange.valueOf() + /* 7 days in milliseconds */ 86400000 * 7);
+
+			if (lastChangePlusOneWeek < new Date()) {
+				vscode.window.showInformationMessage<vscode.MessageItem>(
+					localize(
+						'typescript.openTsServerLog.loggingEnabledForExtendedPeriod',
+						'TS Server logging has been enabled for an extended period. Please set `typescript.tsserver.log` to `off` and restart the TS server to disable logging'),
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:984]: For the message, let's explain why having logging enabled matters, for example:

`TS Server logging is currently enabled which may impact performance` 

which could show with the buttons:

- `Disable logging`
- `Don't show again`

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 3]
```diff
@@ -952,6 +957,47 @@ export default class TypeScriptServiceClient extends Disposable implements IType
...
+					{
+						title: localize(
+							'typescript.openTsServerLog.disableAndReloadOption',
+							'Disable logging and restart TS server'),
+					})
+					.then(selection => {
+						if (selection) {
+							return vscode.workspace.getConfiguration().update('typescript.tsserver.log', 'off', true).then(() => {
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:992]: You can double check but I think the server auto restarts when the log level is changed

File: extensions/typescript-language-features/src/typescriptServiceClient.ts [hunk 4]
```diff
@@ -198,6 +198,9 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 				}
 
 				if (!this._configuration.isEqualTo(oldConfiguration)) {
+					if (oldConfiguration.tsServerLogLevel !== this._configuration.tsServerLogLevel) {
+						context.globalState.update('typescript.tsserver.logLevelChanged', new Date());
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/typescriptServiceClient.ts:202]: How about moving this all to its own class that monitors the log level?

**Review by mjbvz — CHANGES_REQUESTED** [review]

**Comment by nrayburn-tech** [comment]
Updated and I think resolves all of the things you mentioned.  I left a few todo comments about some things that I am not sure about.

Is there an easy way to update values inside of the extension context global state?  Is it stored in a json file that I can open and set the value in, or is it just easier to write something and use the vscode api?

File: extensions/typescript-language-features/src/typeScriptServiceClientHost.ts [hunk 5]
```diff
@@ -147,6 +147,9 @@ export default class TypeScriptServiceClientHost extends Disposable {
 
 		vscode.workspace.onDidChangeConfiguration(this.configurationChanged, this, this._disposables);
 		this.configurationChanged();
+
+		// TODO: not sure if this is the right place to register this.
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/typeScriptServiceClientHost.ts:151]: Yes this looks like a fine location for this check

**Review by mjbvz — CHANGES_REQUESTED** [review]
Looking good overall, I just have a few minor suggestions

File: extensions/typescript-language-features/src/utils/logLevelMonitor.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+	private get doNotPrompt(): boolean {
+		return this.context.globalState.get<boolean | undefined>(LogLevelMonitor.doNotPromptLogLevelStorageKey) || false;
+	}
+
+	private shouldNotifyExtendedLogging(): boolean {
+		// TODO: Should the fallback for this be 0, or new Date()?
+		// 0 Means every user that has logging enabled currently will be notified.
+		// new Date() means that users will not be notified until they have changed their
+		// logging setting at least once.
```
Comment (nrayburn-tech) [inline extensions/typescript-language-features/src/utils/logLevelMonitor.ts:63]: This is the last thing I was unsure about.

**Comment by nrayburn-tech** [comment]
Updated with the changes.  I just have the one last question, commented inline.

File: extensions/typescript-language-features/src/utils/logLevelMonitor.ts (hunk 6 above)
Comment (mjbvz) [reply extensions/typescript-language-features/src/utils/logLevelMonitor.ts:63]: Let's use zero to notify existing users who have logging enabled but don't have a value for `LogLevelMonitor.logLevelChangedStorageKey`

**Review by mjbvz — APPROVED** [review]
Looks good, just remove the todo comment and I'll merge this!

**Comment by mjbvz** [comment]
Thanks! Great work on this PR

This will be in the next VS Code insiders build and is scheduled to be released with VS Code 1.57

**Pull request merged**